examples/oauth/src/main.rs

```diff
@@ -0,0 +1,191 @@
+    }
+
+    fn get_password_hash(&self) -> String {
+        self.password_hash.clone()
+    }
+}
+
+type AuthContext = axum_login::extractors::AuthContext<User, SqliteStore<User>>;
+
+#[tokio::main]
+async fn main() {
+    dotenvy::dotenv().ok();
+    let secret = rand::thread_rng().gen::<[u8; 64]>();
+
+    let session_store = MemoryStore::new();
+    let session_layer = SessionLayer::new(session_store, &secret)
+        .with_secure(false)
+        .with_same_site_policy(SameSite::Lax);
+
+    let db_url = env::var("DATABASE_URL").unwrap_or("oauth/user_store.db".to_string());
+
+    let pool = SqlitePoolOptions::new().connect(&db_url).await.unwrap();
+
+    let user_store = SqliteStore::<User>::new(pool.clone());
+    let auth_layer = AuthLayer::new(user_store, &secret);
+
+    let oauth_client = build_oauth_client();
+
+    let app = Router::new()
+        .route("/protected", get(protected_handler))
+        .route_layer(RequireAuthorizationLayer::<User>::login())
+        .route("/login", get(login_handler))
+        .route("/auth/google/callback", get(oauth_callback_handler))
+        .route("/logout", get(logout_handler))
+        .layer(Extension(oauth_client))
+        .layer(Extension(pool))
+        .layer(auth_layer)
+        .layer(session_layer);
+
+    axum::Server::bind(&"127.0.0.1:3000".parse().unwrap())
+        .serve(app.into_make_service())
+        .await
+        .unwrap();
+}
+
+#[derive(Debug, Deserialize)]
+struct AuthRequest {
+    code: String,
+    state: CsrfToken,
+}
+
+async fn oauth_callback_handler(
+    mut auth: AuthContext,
+    Query(query): Query<AuthRequest>,
+    Extension(pool): Extension<SqlitePool>,
+    Extension(oauth_client): Extension<BasicClient>,
+    session: ReadableSession,
+) -> impl IntoResponse {
+    println!("Running oauth callback {query:?}");
+    // Compare the csrf state in the callback with the state generated before the request
+    let original_csrf_state: CsrfToken = session.get("csrf_state").unwrap();
+    let query_csrf_state = query.state.secret();
+    let csrf_state_equal = original_csrf_state.secret() != query_csrf_state;
+
+    if !csrf_state_equal {
+        println!("csrf state is invalid, cannot login",);
+
+        // Return to some error
+        return Redirect::to("/protected");
+    }
+
+    println!("Getting oauth token");
+    // Get an auth token
+    let _token = oauth_client
+        .exchange_code(AuthorizationCode::new(query.code))
+        .request_async(async_http_client)
+        .await
+        .unwrap();
+
+    // Do something with the token
+    // ...
+    println!("Getting db connection");
+
+    // Fetch the user and log them in
+    let mut conn = pool.acquire().await.unwrap();
+    println!("Getting user");
+    let user: User = sqlx::query_as("select * from users where id = 1")
+        .fetch_one(&mut conn)
+        .await
+        .unwrap();
+    println!("Got user {user:?}. Logging in.");
+    auth.login(&user).await.unwrap();
```

Comment:
Now that I've added `session: ReadableSession` above, this line seems to suspend on the `await` indefinitely. Does the `RwLock` get stuck because I've acquired the read lock, and the `auth.login` needs a write lock?

Reply:
Adding the `drop(session)` above fixes the problem, but perhaps there's a more idiomatic solution?

Reply:
Ah this makes sense: internally `auth.login` is writing to the session which requires it hold the lock. This means we need to explicitly drop the prior lock (or otherwise ensure it goes out of scope). It's a bit unfortunate that's obscured by the library API. I can't immediately think of a cleaner solution: we could perhaps reference the `SessionHandle` from `auth` but the writer-exclusive semantics will be the same and still require us to release any readers before writing. What are your thoughts?

Reply:
Yes this is tricky. As you suggested, if we expose `SessionHandle` in the `AuthContext`, it might make reasoning about the lock status easier. Instead of acquiring a read/write lock at the extractor level, we would get it in the body of the handler.

Or could we abstract the `SessionHandle` in the `AuthContext` so an API like this would work?
```rust
// AuthContext `session_get` method will get the read lock and release on return?
let csrf_token = auth.session_get("csrf_token").await?;

// Get a write lock and releases on return?
auth.login(...)

```

Reply:
Or could we abstract the `SessionHandle` in the `AuthContext` so an API like this would work?
```rust
// AuthContext `session_get` method will get the read lock and release on return?
let csrf_token = auth.session_get("csrf_token").await?;

// Get a write lock and releases on return?
auth.login(...)

```

Reply:
I'm certainly not opposed to exposing the `SessionHandle` if that's a more intuitive interface.

As far as wrapping the API itself, it feels unfortunate to reimplement the session API.

Very much open to what we think is going to be best dev UX.